FWNX-398 Fixing modal dialogs so parents and children behave correctly

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Form.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Form.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Form.cs
@@ -103,6 +103,10 @@
 		private Rectangle		restore_bounds;
 		private bool			autoscale_base_size_set;
 #endif
+
+		internal ArrayList disabled_by_showdialog = new ArrayList();
+		internal static ArrayList modal_dialogs = new ArrayList();
+		
 		#endregion	// Local Variables
 
 		#region Private & Internal Methods
@@ -1787,6 +1791,16 @@
 				XplatUI.UngrabWindow(capture_window);
 			}
 
+			foreach (Form form in Application.OpenForms)
+			{
+				if (form.Enabled == true)
+				{
+					disabled_by_showdialog.Add(form);
+					form.Enabled = false;
+				}
+			}
+			modal_dialogs.Add(this);
+
 #if not
 			// Commented out; we instead let the Visible=true inside the runloop create the control
 			// otherwise setting DialogResult inside any of the events that are triggered by the
@@ -2075,6 +2089,9 @@
 		[EditorBrowsable(EditorBrowsableState.Advanced)]
 		protected virtual void OnActivated(EventArgs e)
 		{
+//			if (!this.Enabled)
+//				return; // prevent Activating of disabled form.
+
 			EventHandler eh = (EventHandler)(Events [ActivatedEvent]);
 			if (eh != null)
 				eh (this, e);
@@ -2660,6 +2677,9 @@
 
 		private void WmClose (ref Message m)
 		{
+			if (this.Enabled == false)
+				return; // prevent closing a disabled form.
+
 			Form act = Form.ActiveForm;
 			// Don't close this form if there's another modal form visible.
 			if (act != null && act != this && act.Modal == true) {
@@ -2744,6 +2764,12 @@
 #endif	
 		private void WmActivate (ref Message m)
 		{
+			if (!this.Enabled && modal_dialogs.Count > 0)
+			{
+				(modal_dialogs[modal_dialogs.Count -1] as Form).Activate ();
+				return; // prevent Activating of disabled form.
+			}
+
 			if (m.WParam != (IntPtr)WindowActiveFlags.WA_INACTIVE) {
 				if (is_loaded) {
 					SelectActiveControl ();
@@ -3241,6 +3267,14 @@
 			FormClosedEventHandler eh = (FormClosedEventHandler)(Events[FormClosedEvent]);
 			if (eh != null)
 				eh (this, e);
+
+			foreach (Form form in disabled_by_showdialog)
+			{
+				form.Enabled = true;
+			}
+			disabled_by_showdialog.Clear();
+			if (modal_dialogs.Contains(this))
+				modal_dialogs.Remove(this);
 		}
 		
 		// Consider calling FireClosingEvents instead of calling this directly.
